src/Core/SettingsChangesHistory.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include <Core/Field.h>
+#include <Core/Settings.h>
+#include <IO/ReadHelpers.h>
+#include <boost/algorithm/string.hpp>
+#include <unordered_map>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+class ClickHouseVersion
+{
+public:
+    ClickHouseVersion(const String & version)
+    {
+        Strings split;
+        boost::split(split, version, [](char c){ return c == '.'; });
+        components.reserve(split.size());
+        if (split.empty())
+            throw Exception{ErrorCodes::BAD_ARGUMENTS, "Cannot parse ClickHouse version here: {}", version};
+
+        for (const auto & split_element : split)
+        {
+            size_t component;
+            if (!tryParse(component, split_element))
+                throw Exception{ErrorCodes::BAD_ARGUMENTS, "Cannot parse ClickHouse version here: {}", version};
+            components.push_back(component);
+        }
+    }
+
+    ClickHouseVersion(const char * version) : ClickHouseVersion(String(version)) {}
+
+    String toString() const
+    {
+        String version = std::to_string(components[0]);
+        for (size_t i = 1; i < components.size(); ++i)
+            version += "." + std::to_string(components[i]);
+
+        return version;
+    }
+
+    bool operator<(const ClickHouseVersion & other) const
+    {
+        return components < other.components;
+    }
+
+    bool operator>=(const ClickHouseVersion & other) const
+    {
+        return components >= other.components;
+    }
+
+private:
+    std::vector<size_t> components;
+};
+
+namespace SettingsChangesHistory
+{
+    struct SettingChange
+    {
+        String name;
+        Field previous_value;
+        Field new_value;
+        String reason;
+    };
+
+    using SettingsChanges = std::vector<SettingChange>;
+}
+
+/// History of settings changes that controls some backward incompatible changes
+/// across all ClickHouse versions. It maps ClickHouse version to settings changes that were done
+/// in this version. Settings changes is a vector of structs {setting_name, previous_value, new_value}
+/// It's used to implement `compatibility` setting (see https://github.com/ClickHouse/ClickHouse/issues/35972)
+std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> settings_changes_history =
+{
+        {"22.7", {{"enable_positional_arguments", false, true, "Enable positional arguments feature by default"}}},
```

Comment:
https://github.com/ClickHouse/ClickHouse/pull/39326
```suggestion
        {"22.7", {{"cross_to_inner_join_rewrite", 1, 2, "Force rewrite comma join to inner"}}},
        {"22.7", {{"enable_positional_arguments", false, true, "Enable positional arguments feature by default"}}},
```